File: April 2024 GFG SOLUTION/April-03.java
//{ Driver Code Starts
import java.io.*;
import java.math.*;
import java.util.*;

class Node {
    int data;
    Node left, right;

    public Node(int d) {
        data = d;
        left = right = null;
    }
}

public class GFG {
    static Node buildTree(String str) {
        // Corner Case
        if (str.length() == 0 || str.equals('N')) return null;
        String[] s = str.split(" ");

        Node root = new Node(Integer.parseInt(s[0]));
        Queue<Node> q = new LinkedList<Node>();
        q.add(root);

        // Starting from the second element
        int i = 1;
        while (!q.isEmpty() && i < s.length) {
            // Get and remove the front of the queue
            Node currNode = q.remove();

            // Get the curr node's value from the string
            String currVal = s[i];

            // If the left child is not null
            if (!currVal.equals("N")) {

                // Create the left child for the curr node
                currNode.left = new Node(Integer.parseInt(currVal));

                // Push it to the queue
                q.add(currNode.left);
            }

            // For the right child
            i++;
            if (i >= s.length) break;
            currVal = s[i];

            // If the right child is not null
            if (!currVal.equals("N")) {

                // Create the right child for the curr node
                currNode.right = new Node(Integer.parseInt(currVal));

                // Push it to the queue
                q.add(currNode.right);
            }

            i++;
        }

        return root;
    }

    public static void main(String args[]) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());
        while (t > 0) {
            String s = br.readLine().trim();
            Node root = buildTree(s);

            String[] snums = br.readLine().split(" ");
            int k = Integer.parseInt(snums[0]);
            int x = Integer.parseInt(snums[1]);
            int y = Integer.parseInt(snums[2]);

            Solution T = new Solution();
            // Call the function with the needed parameters
            System.out.println(T.kthCommonAncestor(root, k, x, y));

            t--;
        }
    }
}

// } Driver Code Ends


// User function Template for Java
class Solution {

    public int kthCommonAncestor(Node root, int k, int x, int y) {
        Node lca = findLowestCommonAncestor(root, x, y);
        ArrayList<Integer> path = new ArrayList<>();
        findPathToNode(root, lca, path);
        Collections.reverse(path);
        if (path.size() < k) return -1;
        return path.get(k - 1);
    }

    Node findLowestCommonAncestor(Node root, int x, int y) {
        if (root == null) return null;
        if (x < root.data && y < root.data) return findLowestCommonAncestor(root.left, x, y);
        if (x > root.data && y > root.data) return findLowestCommonAncestor(root.right, x, y);
        return root;
    }

    void findPathToNode(Node root, Node node, ArrayList<Integer> path) {
        path.add(root.data);
        if (root.data == node.data) return;
        else if (node.data > root.data) findPathToNode(root.right, node, path);
        else findPathToNode(root.left, node, path);
    }
}

File: April 2024 GFG SOLUTION/April-03.py
#User function Template for python3

'''
class Node:
    def _init_(self, data):
        self.data = data
        self.left = None
        self.right = None
'''

class Solution:
    def findLowestCommonAncestor(self, root, x, y):
        if root is None:
            return None
        if x < root.data and y < root.data:
            return self.findLowestCommonAncestor(root.left, x, y)
        if x > root.data and y > root.data:
            return self.findLowestCommonAncestor(root.right, x, y)
        return root
    
    def findPathToNode(self, root, node, path):
        path.append(root.data)
        if root.data == node.data:
            return
        elif node.data > root.data:
            self.findPathToNode(root.right, node, path)
        else:
            self.findPathToNode(root.left, node, path)
    
    def kthCommonAncestor(self, root, k, x, y):
        lca = self.findLowestCommonAncestor(root, x, y)
        path = []
        self.findPathToNode(root, lca, path)
        path.reverse()
        if len(path) < k:
            return -1
        return path[k - 1]


 # Driver Code Starts

#Initial Template for Python 3


from collections import deque

class Node:
    def _init_(self, data):
        self.data = data
        self.left = None
        self.right = None

# Function to Build Tree
def buildTree(str):
    # Corner Case
    if len(str) == 0 or str[0] == 'N':
        return None

    # Creating list of strings from input string after splitting by space
    ip = str.split()

    # Create the root of the tree
    root = Node(int(ip[0]))

    # Push the root to the queue
    queue = deque()
    queue.append(root)

    # Starting from the second element
    i = 1
    while queue and i < len(ip):
        # Get and remove the front of the queue
        currNode = queue.popleft()

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if currVal != "N":
            # Create the left child for the current node
            currNode.left = Node(int(currVal))

            # Push it to the queue
            queue.append(currNode.left)

        # For the right child
        i += 1
        if i >= len(ip):
            break
        currVal = ip[i]

        # If the right child is not null
        if currVal != "N":
            # Create the right child for the current node
            currNode.right = Node(int(currVal))

            # Push it to the queue
            queue.append(currNode.right)

        i += 1

    return root

for _ in range(int(input())):
    s = input()
    root = buildTree(s)
    k, x, y = map(int, input().split())
    if root is None:
        continue
    
    if root.left is None and root.right is None:
        continue
    
    ob = Solution()
    print(ob.kthCommonAncestor(root, k, x, y))


# } Driver Code Ends
